File: tools/build-standalone.js
#!/usr/bin/env node
/* ===========================================================
   يعيد بناء standalone.html من ملفات assets/
   -----------------------------------------------------------
   standalone.html يحتوي على الخطوط والشعار مضمّنة (base64) وهي
   غير موجودة في المستودع، لذلك لا نبنيه من الصفر: نستبدل فقط
   محتوى وسوم <script> الثلاثة بمحتوى config.js و data.js و app.js،
   مع تحويل مسار الشعار إلى data URI.

   وكذلك نستبدل ملفَّي CSS داخل وسم <style>، بين علامتَي تعليق
   يحملان @@css و @@end. ما قبل الأولى (الخطوط المضمّنة) وما بعد
   الثانية يبقى كما هو.

   التشغيل:  node tools/build-standalone.js
   =========================================================== */
const fs = require("fs");
const path = require("path");

const ROOT = path.join(__dirname, "..");
const TARGET = path.join(ROOT, "standalone.html");
const SOURCES = ["assets/config.js", "assets/data.js", "assets/app.js"];
const LOGO = "assets/adaa-logo.png";

const html = fs.readFileSync(TARGET, "utf8");
const lines = html.split("\n");

// حدود وسوم <script> المستقلة (سطر يحتوي الوسم وحده)
const opens = [];
const closes = [];
lines.forEach((l, i) => {
  if (l.trim() === "<script>") opens.push(i);
  if (l.trim() === "</script>") closes.push(i);
});

if (opens.length !== SOURCES.length || closes.length !== SOURCES.length) {
  console.error(`✗ توقعت ${SOURCES.length} وسوم <script> لكن وجدت ${opens.length}. أوقفت البناء.`);
  process.exit(1);
}

const logoUri = "data:image/png;base64," + fs.readFileSync(path.join(ROOT, LOGO)).toString("base64");

// نبني من الآخر للأول حتى لا تتغيّر أرقام الأسطر أثناء الاستبدال
const out = lines.slice();
for (let i = SOURCES.length - 1; i >= 0; i--) {
  let code = fs.readFileSync(path.join(ROOT, SOURCES[i]), "utf8").replace(/\n+$/, "");
  code = code.split(LOGO).join(logoUri);
  out.splice(opens[i] + 1, closes[i] - opens[i] - 1, code);
}

let result = out.join("\n");

// الأنماط: نستبدل ما بين العلامتين فقط، ونترك الخطوط المضمّنة قبلهما
const CSS_SOURCES = ["assets/base.css", "assets/app.css"];
const css = CSS_SOURCES
  .map((f) => fs.readFileSync(path.join(ROOT, f), "utf8").trim())
  .join("\n");
const marked = /\/\* @@css @@ \*\/\n[\s\S]*?\n\/\* @@end @@ \*\//;
if (!marked.test(result)) {
  console.error("✗ لم أجد علامتي الأنماط (@@css / @@end) داخل <style>. أوقفت البناء.");
  process.exit(1);
}
result = result.replace(marked, `/* @@css @@ */\n${css}\n/* @@end @@ */`);

fs.writeFileSync(TARGET, result);

if (result.includes(LOGO)) {
  console.error("✗ بقي مسار شعار غير مضمّن في الملف الناتج.");
  process.exit(1);
}

console.log(`✓ حُدِّث standalone.html من: ${SOURCES.join("، ")}`);

File: build-standalone.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
يعيد بناء standalone.html من ملفات assets/.

standalone.html نسخة بملف واحد من التطبيق: تضمّ الخطوط والشعار مُرمّزة داخلها،
إضافة إلى أنماط وأسطر خاصة بها وحدها. المناطق المنسوخة من assets/ محاطة
بعلامات في الملف:

    /* @@asset app.js @@ */  …المحتوى…  /* @@end @@ */

هذا السكربت يستبدل ما بين كل زوج علامات بمحتوى الملف الحالي في assets/،
ولا يمسّ أي شيء خارجها. شغّله بعد أي تعديل على assets/:

    python3 build-standalone.py
"""
import io
import os
import re
import sys

ROOT = os.path.dirname(os.path.abspath(__file__))
TARGET = os.path.join(ROOT, "standalone.html")
LOGO_PATH = "assets/adaa-logo.png"
MARKED = re.compile(r"/\* @@asset (\S+) @@ \*/\n.*?\n/\* @@end @@ \*/", re.S)


def read(*parts):
    with io.open(os.path.join(ROOT, *parts), encoding="utf-8") as fh:
        return fh.read()


def main():
    html = read("standalone.html")

    # الشعار يُضمَّن داخل standalone كـ data URI بدل مسار الملف
    logo = re.search(r'src="(data:image/png;base64,[^"]+)"', html)
    if not logo:
        sys.exit("لم يُعثر على الشعار المضمّن (data:image/png) في standalone.html")
    logo_uri = logo.group(1)

    seen = []

    def swap(m):
        name = m.group(1)
        path = os.path.join(ROOT, "assets", name)
        if not os.path.exists(path):
            sys.exit("العلامة تشير إلى ملف غير موجود: assets/%s" % name)
        body = read("assets", name).strip().replace(LOGO_PATH, logo_uri)
        seen.append(name)
        return "/* @@asset %s @@ */\n%s\n/* @@end @@ */" % (name, body)

    html, n = MARKED.subn(swap, html)
    if not n:
        sys.exit("لم يُعثر على أي علامات @@asset في standalone.html")

    with io.open(TARGET, "w", encoding="utf-8") as fh:
        fh.write(html)
    print("حُدّثت %d مناطق (%s) — standalone.html: %.1f كيلوبايت"
          % (n, "، ".join(seen), os.path.getsize(TARGET) / 1024.0))


if __name__ == "__main__":
    main()
